mm2src/lp_ordermatch.rs

```diff
@@ -3285,6 +3361,146 @@ pub async fn set_price(ctx: MmArc, req: Json) -> Result<Response<Vec<u8>>, Strin
+            try_s!(validate_price(new_price.clone()));
+            update_msg = update_msg.with_new_price(new_price.clone().into());
+            new_price
+        },
+        None => original_price,
+    };
+
+    let min_base_amount = base_coin.min_trading_vol();
+    let min_rel_amount = rel_coin.min_trading_vol();
+
+    // Add min_volume to update_msg if min_volume is found in the request
+    if let Some(min_volume) = req.min_volume.clone() {
+        // Validate and Ca lculate Minimum Volume
+        let actual_min_vol = try_s!(validate_and_get_min_vol(
+            min_base_amount.clone(),
+            min_rel_amount.clone(),
+            Some(min_volume),
+            new_price.clone()
+        ));
+        update_msg = update_msg.with_new_min_volume(actual_min_vol.into());
+    }
+
+    // Calculate order volume and add to update_msg if new_volume is found in the request
+    let new_volume = if req.max.unwrap_or(false) {
+        let max_volume = try_s!(get_max_volume(&ctx, &base_coin, &rel_coin).await) + reserved_amount;
+        update_msg = update_msg.with_new_max_volume(max_volume.clone().into());
+        max_volume
+    } else if Option::is_some(&req.volume_delta) {
+        let volume = original_volume + req.volume_delta.unwrap();
+        if volume <= MmNumber::from("0") {
+            return ERR!("New volume {} should be more than zero", volume);
+        }
+        try_s!(
+            check_balance_for_maker_swap(
+                &ctx,
+                &base_coin,
+                &rel_coin,
+                volume.clone(),
+                None,
+                None,
+                FeeApproxStage::OrderIssue
+            )
+            .await
+        );
+        update_msg = update_msg.with_new_max_volume(volume.clone().into());
+        volume
+    } else {
+        original_volume
+    };
+
+    // Validate Order Volume
+    try_s!(validate_max_vol(
+        min_base_amount.clone(),
+        min_rel_amount.clone(),
+        new_volume.clone(),
+        req.min_volume.clone(),
+        new_price
+    ));
+
+    let mut my_maker_orders = ordermatch_ctx.my_maker_orders.lock().await;
+    let (rpc_result, base, rel) = match my_maker_orders.get_mut(&req.uuid) {
+        None => return ERR!("Order with UUID: {} has been deleted", req.uuid),
+        Some(order) => {
+            if order.matches.len() != matches.len() || !order.matches.keys().all(|k| matches.contains_key(k)) {
+                return ERR!("Order {} is being matched now, can't update", req.uuid);
+            }
+            let reserved_amount = order.reserved_amount();
+            if new_volume <= reserved_amount {
+                return ERR!(
+                    "New volume {} should be more than reserved amount for order matches {}",
+                    new_volume,
+                    reserved_amount
+                );
+            }
+            update_msg = update_msg.with_new_max_volume((new_volume - order.reserved_amount()).into());
```

Comment:
This results in invalid volume being set for order:
1. Let's say the current `max_base_vol = 5`, `reserved = 3` so `available = 2`.
2. We set `volume_delta = 2` in the request. The `new_volume = original_volume + volume_delta = 7`.
3. Then we deduct `new_volume - reserved_amount` so `max_base_vol = 7 - 3 = 4`, which is even lower than initial one.

You also have `let reserved_amount = order.reserved_amount();` in a couple of lines above, you can reuse it to avoid calculation duplication overhead.

Reply:
yeah I had it after `order.apply_updated(&update_msg)` but must have messed it up later. Will fix it.

Reply:
I can also use `reserved_amount` from here https://github.com/KomodoPlatform/atomicDEX-API/blob/db9c095995cc586c65573a68eada05752189f95c/mm2src/lp_ordermatch.rs#L3370
avoiding calculation in `let reserved_amount = order.reserved_amount();` too, since we made sure It won't change through the update call.